dupemap.1
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.14
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "DUPEMAP 1"
.TH DUPEMAP 1 "2004-04-30" "1.1.1" "Magic Rescue"
.SH "NAME"
dupemap \- Creates a database of file checksums and uses it to eliminate
duplicates
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBdupemap\fR [ \fIoptions\fR ] [ \fB\-d\fR \fIdatabase\fR ] \fIoperation\fR \fIpath...\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBdupemap\fR recursively scans each \fIpath\fR to find checksums of file contents.
Directories are searched through in no particular order.  Its actions depend on
whether the \fB\-d\fR option is given, and on the \fIoperation\fR parameter, which
must be a comma-seperated list of \fBscan\fR, \fBreport\fR, \fBdelete\fR:
.Sh "Without \fB\-d\fP"
.IX Subsection "Without -d"
\&\fBdupemap\fR will take action when it sees the same checksum repeated more than
once, i.e. it simply finds duplicates recursively.  The action depends on
\&\fIoperation\fR:
.IP "\fBreport\fR" 7
.IX Item "report"
Report what files are encountered more than once, printing their names to
standard output.
.IP "\fBdelete\fR[\fB,report\fR]" 7
.IX Item "delete[,report]"
Delete files that are encountered more than once.  Print their names if
\&\fBreport\fR is also given.
.Sh "With \fB\-d\fP"
.IX Subsection "With -d"
The \fIdatabase\fR argument to \fB\-d\fR will denote a database file (see the
\&\*(L"\s-1DATABASE\s0\*(R" section in this manual for details) to read from or write to.  In
this mode, the \fBscan\fR operation should be run on one \fIpath\fR, followed by the
\&\fBreport\fR or \fBdelete\fR operation on another (\fInot the same!\fR) \fIpath\fR.
.IP "\fBscan\fR" 7
.IX Item "scan"
Add the checksum of each file to \fIdatabase\fR.  This operation must be run
initially to create the database.  To start over, you must manually delete the
database file(s) (see the \*(L"\s-1DATABASE\s0\*(R" section).
.IP "\fBreport\fR" 7
.IX Item "report"
Print each file name if its checksum is found in \fIdatabase\fR.
.IP "\fBdelete\fR[\fB,report\fR]" 7
.IX Item "delete[,report]"
Delete each file if its checksum is found in \fIdatabase\fR.  If \fBreport\fR is also
present, print the name of each deleted file.
.Sp
\&\fI\s-1WARNING:\s0\fR use the \fBreport\fR operation first to see what will be deleted.
.Sp
\&\fI\s-1WARNING:\s0\fR if you run \fBdupemap delete\fR on the same \fIpath\fR you just ran
\&\fBdupemap scan\fR on, it will \fIdelete every file!\fR The idea of these options is
to scan one \fIpath\fR and delete files in a second \fIpath\fR.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-d\fR \fIdatabase\fR" 7
.IX Item "-d database"
Use \fIdatabase\fR as an on-disk database to read from or write to.  See the
\&\*(L"\s-1DESCRIPTION\s0\*(R" section above about how this influences the operation of
\&\fBdupemap\fR.
.IP "\fB\-m\fR \fIminsize\fR" 7
.IX Item "-m minsize"
Ignore files below this size.
.IP "\fB\-M\fR \fImaxsize\fR" 7
.IX Item "-M maxsize"
Ignore files above this size.
.SH "USAGE"
.IX Header "USAGE"
.Sh "General usage"
.IX Subsection "General usage"
The easiest operations to understand is when the \fB\-d\fR option is not given.  To
delete all duplicate files in \fI/tmp/recovered\-files\fR, do:
.PP
.Vb 1
\&    $ dupemap delete /tmp/recovered files
.Ve
.PP
Often, \fBdupemap scan\fR is run to produce a checksum database of all files in a
directory tree.  Then \fBdupemap delete\fR is run on another directory, possibly 
following \fBdupemap report\fR.  For example, to delete all files in
\&\fI/tmp/recovered\-files\fR that already exist in \fI$HOME\fR, do this:
.PP
.Vb 2
\&    $ dupemap -d homedir.map scan $HOME
\&    $ dupemap -d homedir.map delete,report /tmp/recovered-files
.Ve
.Sh "Usage with magicrescue"
.IX Subsection "Usage with magicrescue"
The main application for \fBdupemap\fR is to take some pain out of performing
undelete operations with \fBmagicrescue\fR(1).  The reason is that \fBmagicrescue\fR
will extract every single file of the specified type on the block device, so
undeleting files requires you to find a few files out of hundreds, which can
take a long time if done manually.  What we want to do is to only extract the
documents that don't exist on the file system already.
.PP
In the following scenario, you have accidentally deleted some important Word
documents in Windows.  If this were a real-world scenario, then by all means use
The Sleuth Kit.  However, \fBmagicrescue\fR will work even when the directory
entries were overwritten, i.e. more files were stored in the same folder later.
.PP
You boot into Linux and change to a directory with lots of space.  Mount the
Windows partition, preferably read-only (especially with \s-1NTFS\s0), and create the
directories we will use.
.PP
.Vb 2
\&    $ mount -o ro /dev/hda1 /mnt/windows
\&    $ mkdir healthy_docs rescued_docs
.Ve
.PP
Extract all the healthy Word documents with \fBmagicrescue\fR and build a database
of their checksums.  It may seem a little redundant to send all the documents
through \fBmagicrescue\fR first, but the reason is that this process may modify
them (e.g. stripping trailing garbage), and therefore their checksum will not
be the same as the original documents.  That is quite rare with Word documents,
but this step is critical for formats like mp3, gzip and jpeg, which will often
look very different after passing through magicrescue.
.PP
.Vb 4
\&    $ find /mnt/windows -type f -iname "*.doc" -print0 \e
\&      |xargs -0 magicrescue -r msoffice -d healthy_docs
\&    $ dupemap -d healthy_docs.map scan healthy_docs
\&    $ rm -rf healthy_docs
.Ve
.PP
This can be done more efficiently, albeit with a more complicated command line.
The following pipeline will do the same thing as the commands above, but it's
faster (especially for multi-CPU machines) and will use very little temporary
disk space.
.PP
.Vb 4
\&    $ find /mnt/windows -type f -iname "*.doc" -print0 \e
\&      |xargs -0 magicrescue -Mo -r msoffice -d healthy_docs \e
\&      |sed "s/['\e"\e\e[:blank:]]/\e\e\e\e&/g" \e
\&      |xargs sh -c 'dupemap -d healthy_docs.map scan "$@"; rm -f "$@"' sh
.Ve
.PP
Now rescue all \f(CW\*(C`msoffice\*(C'\fR documents from the block device and get rid of
everything that's not a *.doc.
.PP
.Vb 2
\&    $ magicrescue -Mo -r msoffice -d rescued_docs /dev/hda1 \e
\&      |grep -v '.doc$'|xargs rm -f
.Ve
.PP
Remove all the rescued documents that also appear on the file system, and
remove duplicates.
.PP
.Vb 2
\&    $ dupemap -d healthy_docs.map delete,report rescued_docs
\&    $ dupemap delete,report rescued_docs
.Ve
.PP
The \fIrescued_docs\fR folder should now contain only a few files.  This will be
the undeleted files and some documents that were not stored in contiguous
blocks (use that defragger ;\-)).
.PP
For the record, the two commands for rescueing files and removing duplicates
can also be expressed with one pipeline.
.PP
.Vb 4
\&    $ magicrescue -Mo -r msoffice -d rescued_docs /dev/hda1 \e
\&      |xargs sh -c 'for f in "$@"; do echo "$f"; done|grep -v "\e.doc$" \e
\&             |xargs rm -f; \e
\&             dupemap -v delete healthy_docs.map "$@" 2>/dev/null' sh
.Ve
.Sh "Usage with fsck"
.IX Subsection "Usage with fsck"
In this scenario (based on a true story), you have a hard disk that's gone bad.
You have managed to \fIdd\fR about 80% of the contents into the file \fIdiskimage\fR,
and you have an old backup from a few months ago.  The disk is using reiserfs
on Linux.
.PP
First, use fsck to make the file system usable again.  It will find many
nameless files and put them in \fIlost+found\fR.  You need to make sure there is
some free space on the disk image, so fsck has something to work with.
.PP
.Vb 6
\&    $ cp diskimage diskimage.bak
\&    $ dd if=/dev/zero bs=1M count=2048 >> diskimage
\&    $ reiserfsck --rebuild-tree diskimage
\&    $ mount -o loop diskimage /mnt
\&    $ ls /mnt/lost+found
\&    (tons of files)
.Ve
.PP
Our strategy will be to restore the system with the old backup as a base and
merge the two other sets of files (\fI/mnt/lost+found\fR and \fI/mnt\fR) into the
backup after eliminating duplicates.  Therefore we create a checksum database
of the directory we have unpacked the backup in.
.PP
.Vb 1
\&    $ dupemap -d backup.map scan ~/backup
.Ve
.PP
Next, we eliminate all the files from the rescued image that are also present
in the backup.
.PP
.Vb 1
\&    $ dupemap -d backup.map delete,report /mnt
.Ve
.PP
We also want to remove duplicates from \fIlost+found\fR, and we want to get rid of

NEWS
Version 1.1.9: Mon Apr 12 20:21:16 CEST 2010
- Recipe for rar archives

Version 1.1.8: Wed Nov  4 16:12:53 CET 2009
- Add mbox recipes, contributed by Jeff Bucove.
- Raise maximum virtual memory usage for helper programs from 50MB to 1024MB.
  It turns out that mencoder needs around 150MB on 64-bit systems.
- Place tools directory first in path to avoid name clashes with other
  programs.

Version 1.1.7: Fri Sep  4 19:48:14 CEST 2009
- Update zip recipe to work with latest version of Zip and latest
  OpenOffice.org file formats (thanks to Fernando)

Version 1.1.6: Sun Feb 15 22:13:54 CET 2009
- Recipe for Nikon raw photos
- Recipe for large PPM files

Version 1.1.5: Thu Jan 17 14:03:49 CET 2008
- Recipe for Flac files
- Recipe for Canon CR2 RAW files
- Configure script compatibility fix for Debian-based Linuxes

Version 1.1.4: Sun Aug 15 01:11:27 CEST 2004
- Fixed a minor bug that prevented some not yet written recipes from working
  perfectly

Version 1.1.3: Tue Aug 10 22:32:42 CEST 2004
- Fixed a minor bug to make jpeg matching 100% reliable instead of 99.4%
- Added GIMP xcf recipe
- Added text extraction and experimental recipes for perl and gpl
- Added test suite to ensure correctness
- Support for dupemap on Debian and others
- Extraction of msoffice on cygwin

Version 1.1.2: Thu May  6 22:47:52 CEST 2004
- Added the -O option for easy resuming from any offset
- Added the -I option for easier usage with find(1)
- Reorganized source code
- Wrote more documentation

Version 1.1.1: Sat May  1 19:14:14 CEST 2004
- Added recipe for zip files and OpenOffice.org documents
- Added -b option for block alignment
- If argument to -r is a directory, read all files within it
- Anti-overlapping between files is now global instead of per-recipe
- Fixed bug that made anti-overlap code skip too much sometimes
- Fixed bug that prevented the use of an absolute path for recipes
- dupemap works on KNOPPIX now

Version 1.1.0: Tue Apr 13 13:23:28 CEST 2004
- Transformed README files into proper manpages and wrote more documentation
- Added dupemap, another tool for getting a better overview of the output
- Recipes take the output file name in $1 now, instead of $0
- Changed the output file format to adapt to "find|xargs magicrescue" usage
- Make magicrescue aware of output-file renaming
- Ported to Solaris
- Added more choices for machine-readable output

Version 1.0.4: Fri Mar 26 18:17:05 CET 2004
- tools/checkrecipe was broken for some file types, fixed that
- Fixed a rounding bug that could lead to segfaulting on some rare data
- Limits memory to 50MB per process, to prevent helper programs from exhausting
  memory on invalid input

Version 1.0.3: Tue Mar 23 16:02:19 CET 2004
- New recipes for Microsoft Office formats and ELF
- Recipe for gzip now extracts file names, if available
- Proper support was added for resuming an interrupted run
- Output file names now reflect match offsets
- Portability fixes, especially for Cygwin

Version 1.0.2: Mon Mar 15 13:17:48 CET 2004
- Much improved mp3 recipe
- Anti-overlapping prevents the same recipe from extracting part of the same
  file twice
- Added magicsort for categorizing the output files
- Magic Rescue now has a "normal" build system (./configure, make, make install)
- Bugfixes in the buffer overlapping code
- Prevents helper programs from leaving core files

Version 1.0.1: Fri Mar  5 11:15:04 CET 2004
- Nicer memory management
- Improved .avi recipe
- mp3 recipe
- gzip recipe

Version 1.0.0: Mon Mar  1 21:56:45 CET 2004
- Initial public release
- Has recipe parser and comes with recipes for avi, jpeg and png



src/dupemap.c
/*
 * dupemap, part of Magic Rescue
 * Copyright (C) 2004 Jonas Jensen <jbj@knef.dk>
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
#include "config.h"

#ifdef HAVE_NDBM

#include <sys/types.h>
#include <sys/stat.h>

#include <errno.h>
#include <dirent.h>
#include <fcntl.h>
#include <limits.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "find_dbm.h"
#include "recur.h"
#include "util.h"

enum operation { SCAN=1, REPORT=2, DELETE=4, HASDB=8 };

enum mode { MODE_READDB=1, MODE_WRITEDB=2, MODE_DELETE=4, MODE_REPORT=8,
            MODE_TEMPDB=16 };

/*
 *
 * DATABASE CODE
 *
 * An implementation of the database layer should provide a dbhandle typedef
 * and the csdb_{open,open_tempdb,exists,store,close} functions.  No
 * values are stored in the database, only the existence of keys matters.
 *
 */
typedef struct {
    DBM *db;
    char *tempdir;
} dbhandle;

static int csdb_open_tempdb(dbhandle *db)
{
    int i = 0, rv;
    char *env_tmp = getenv("TMP");
    if (!env_tmp || !*env_tmp)
	env_tmp = "/tmp";

    db->tempdir = malloc(PATH_MAX);
    do {
	snprintf(db->tempdir, PATH_MAX, "%s/dupemap_%d", env_tmp, i++);
	rv = mkdir(db->tempdir, 0700);
    } while (rv != 0 && errno == EEXIST);

    if (rv != 0) {
	perror("making temporary directory");
	free(db->tempdir);
	return -1;

    } else {
	char file[PATH_MAX];
	snprintf(file, PATH_MAX, "%s/tempdb", db->tempdir);
	db->db = dbm_open(file, O_RDWR | O_CREAT, 0666);

	if (!db->db) {
	    perror("dbm_open");
	    free(db->tempdir);
	    return -1;
	}
    }

    return 0;
}

static int csdb_open(dbhandle *db, char *file, enum mode mode)
{
    int openmode = mode & MODE_READDB ? O_RDONLY : O_RDWR | O_CREAT;

    db->tempdir = NULL;
    db->db = dbm_open(file, openmode, 0666);
    if (!db->db) {
        perror("dbm_open");
	return -1;
    }
    return 0;
}

static int csdb_exists(dbhandle *db, void *key, int len)
{
    datum d, value;
    d.dptr = key;
    d.dsize = len;

    value = dbm_fetch(db->db, d);
    return value.dptr != NULL;
}

static int csdb_store(dbhandle *db, void *key, int len)
{
    datum d, value;

    d.dptr = key;
    d.dsize = len;
    value.dptr = "";
    value.dsize = DBM_KEY_LEN;

    dbm_store(db->db, d, value, DBM_INSERT);
    return 0;
}

static int csdb_close(dbhandle *db)
{
    dbm_close(db->db);

    if (db->tempdir) {
	rm_rf(db->tempdir);
	free(db->tempdir);
    }
    return 0;
}

/*
 * 
 * CHECKSUM CODE
 *
 * An implementation should provide the checksum_t typedef and the
 * checksum_calc function.
 *
 */

/* This CRC32 implementation is derived from Weeder
 * <http://members.teleweb.at/erikajo/weeder.htm>
 */
static const unsigned long crctab[256] = {
0x0,
0x04C11DB7,0x09823B6E,0x0D4326D9,0x130476DC,0x17C56B6B,0x1A864DB2,0x1E475005,
0x2608EDB8,0x22C9F00F,0x2F8AD6D6,0x2B4BCB61,0x350C9B64,0x31CD86D3,0x3C8EA00A,
0x384FBDBD,0x4C11DB70,0x48D0C6C7,0x4593E01E,0x4152FDA9,0x5F15ADAC,0x5BD4B01B,
0x569796C2,0x52568B75,0x6A1936C8,0x6ED82B7F,0x639B0DA6,0x675A1011,0x791D4014,
0x7DDC5DA3,0x709F7B7A,0x745E66CD,0x9823B6E0,0x9CE2AB57,0x91A18D8E,0x95609039,
0x8B27C03C,0x8FE6DD8B,0x82A5FB52,0x8664E6E5,0xBE2B5B58,0xBAEA46EF,0xB7A96036,
0xB3687D81,0xAD2F2D84,0xA9EE3033,0xA4AD16EA,0xA06C0B5D,0xD4326D90,0xD0F37027,
0xDDB056FE,0xD9714B49,0xC7361B4C,0xC3F706FB,0xCEB42022,0xCA753D95,0xF23A8028,
0xF6FB9D9F,0xFBB8BB46,0xFF79A6F1,0xE13EF6F4,0xE5FFEB43,0xE8BCCD9A,0xEC7DD02D,
0x34867077,0x30476DC0,0x3D044B19,0x39C556AE,0x278206AB,0x23431B1C,0x2E003DC5,
0x2AC12072,0x128E9DCF,0x164F8078,0x1B0CA6A1,0x1FCDBB16,0x018AEB13,0x054BF6A4,
0x0808D07D,0x0CC9CDCA,0x7897AB07,0x7C56B6B0,0x71159069,0x75D48DDE,0x6B93DDDB,
0x6F52C06C,0x6211E6B5,0x66D0FB02,0x5E9F46BF,0x5A5E5B08,0x571D7DD1,0x53DC6066,
0x4D9B3063,0x495A2DD4,0x44190B0D,0x40D816BA,0xACA5C697,0xA864DB20,0xA527FDF9,
0xA1E6E04E,0xBFA1B04B,0xBB60ADFC,0xB6238B25,0xB2E29692,0x8AAD2B2F,0x8E6C3698,
0x832F1041,0x87EE0DF6,0x99A95DF3,0x9D684044,0x902B669D,0x94EA7B2A,0xE0B41DE7,
0xE4750050,0xE9362689,0xEDF73B3E,0xF3B06B3B,0xF771768C,0xFA325055,0xFEF34DE2,
0xC6BCF05F,0xC27DEDE8,0xCF3ECB31,0xCBFFD686,0xD5B88683,0xD1799B34,0xDC3ABDED,
0xD8FBA05A,0x690CE0EE,0x6DCDFD59,0x608EDB80,0x644FC637,0x7A089632,0x7EC98B85,
0x738AAD5C,0x774BB0EB,0x4F040D56,0x4BC510E1,0x46863638,0x42472B8F,0x5C007B8A,
0x58C1663D,0x558240E4,0x51435D53,0x251D3B9E,0x21DC2629,0x2C9F00F0,0x285E1D47,
0x36194D42,0x32D850F5,0x3F9B762C,0x3B5A6B9B,0x0315D626,0x07D4CB91,0x0A97ED48,
0x0E56F0FF,0x1011A0FA,0x14D0BD4D,0x19939B94,0x1D528623,0xF12F560E,0xF5EE4BB9,
0xF8AD6D60,0xFC6C70D7,0xE22B20D2,0xE6EA3D65,0xEBA91BBC,0xEF68060B,0xD727BBB6,
0xD3E6A601,0xDEA580D8,0xDA649D6F,0xC423CD6A,0xC0E2D0DD,0xCDA1F604,0xC960EBB3,
0xBD3E8D7E,0xB9FF90C9,0xB4BCB610,0xB07DABA7,0xAE3AFBA2,0xAAFBE615,0xA7B8C0CC,
0xA379DD7B,0x9B3660C6,0x9FF77D71,0x92B45BA8,0x9675461F,0x8832161A,0x8CF30BAD,
0x81B02D74,0x857130C3,0x5D8A9099,0x594B8D2E,0x5408ABF7,0x50C9B640,0x4E8EE645,
0x4A4FFBF2,0x470CDD2B,0x43CDC09C,0x7B827D21,0x7F436096,0x7200464F,0x76C15BF8,
0x68860BFD,0x6C47164A,0x61043093,0x65C52D24,0x119B4BE9,0x155A565E,0x18197087,
0x1CD86D30,0x029F3D35,0x065E2082,0x0B1D065B,0x0FDC1BEC,0x3793A651,0x3352BBE6,
0x3E119D3F,0x3AD08088,0x2497D08D,0x2056CD3A,0x2D15EBE3,0x29D4F654,0xC5A92679,
0xC1683BCE,0xCC2B1D17,0xC8EA00A0,0xD6AD50A5,0xD26C4D12,0xDF2F6BCB,0xDBEE767C,
0xE3A1CBC1,0xE760D676,0xEA23F0AF,0xEEE2ED18,0xF0A5BD1D,0xF464A0AA,0xF9278673,
0xFDE69BC4,0x89B8FD09,0x8D79E0BE,0x803AC667,0x84FBDBD0,0x9ABC8BD5,0x9E7D9662,
0x933EB0BB,0x97FFAD0C,0xAFB010B1,0xAB710D06,0xA6322BDF,0xA2F33668,0xBCB4666D,
0xB8757BDA,0xB5365D03,0xB1F740B4
};

typedef struct {
    long crc;
    off_t filesize;
} checksum_t;

#define BUFLEN ( 1 << 16 )

static int checksum_calc(checksum_t *checksum, const char *file)
{
    static char buf[BUFLEN];
    ssize_t bytes_read;
    off_t _length = 0;
    long _crc = 0;
    FILE *fp;

    // Clear any padding between the fields
    memset(checksum, 0, sizeof(checksum_t));

    fp = fopen (file, "rb");
    if (fp == NULL) {
	return -1;
    }

    while ((bytes_read = fread(buf, 1, BUFLEN, fp)) > 0) {
	char *cp = buf;

	_length += bytes_read;
	while (bytes_read--)
	    _crc = (_crc << 8) ^ crctab[((_crc >> 24) ^ *(cp++)) & 0xFF];
    }

    if (ferror(fp)) {
	fclose(fp);
	return -1;
    }
    fclose(fp);

    checksum->filesize = _length;

    for (; _length; _length >>=8 )
	_crc = (_crc << 8) ^ crctab[((_crc >> 24) ^ _length) & 0xFF];

    checksum->crc = ~_crc & 0xFFFFFFFF;

    return 0;
}

/*
 *
 * MAIN PROGRAM
 *
 */
static dbhandle *global_dbhandle = NULL;
static off_t min_size = 1, max_size = 0;
static enum mode mode;

static void signal_beforedeath(int signo)
{
    fprintf(stderr, "\ndupemap: killed by signal %d\n", signo);
    csdb_close(global_dbhandle);

    exit(0);
}

static enum mode parse_mode(char *str, int hasdb) {
    char *tok;
    enum operation op = 0;
    unsigned int i;

    struct { char *opname; int opval; } opnames[] = {
	{ "scan",   SCAN    },
	{ "report", REPORT  },
	{ "delete", DELETE  },
    };
    struct { enum operation op; enum mode mode; } op2mode[] = {
	{ HASDB | SCAN,            MODE_WRITEDB },
	{ HASDB | SCAN   | REPORT, MODE_WRITEDB | MODE_REPORT },

	{ HASDB | REPORT,          MODE_READDB | MODE_REPORT },
	{ HASDB | DELETE,          MODE_READDB | MODE_DELETE },
	{ HASDB | DELETE | REPORT, MODE_READDB | MODE_DELETE | MODE_REPORT },

	{ REPORT                , MODE_TEMPDB | MODE_REPORT },
	{ DELETE                , MODE_TEMPDB | MODE_DELETE },
	{ DELETE | REPORT       , MODE_TEMPDB | MODE_DELETE | MODE_REPORT },
	{ SCAN | REPORT         , MODE_TEMPDB | MODE_REPORT },
	{ SCAN | DELETE         , MODE_TEMPDB | MODE_DELETE },
	{ SCAN | DELETE | REPORT, MODE_TEMPDB | MODE_DELETE | MODE_REPORT },
    };

    for (tok = strtok(str, ", "); tok; tok = strtok(NULL, ", ")) {
	for (i = 0; i < sizeof(opnames)/sizeof(*opnames); i++) {
	    if (strcmp(tok, opnames[i].opname) == 0)
		op |= opnames[i].opval;
	}
    }
    if (hasdb) op |= HASDB;
    
    for (i = 0; i < sizeof(op2mode)/sizeof(*op2mode); i++) {
	if (op2mode[i].op == op)
	    return op2mode[i].mode;
    }
    return 0;
}

static void scan_file(const char *name, const struct stat *st,
	dbhandle *db) {
    checksum_t checksum;

    if (!(st->st_mode & S_IFREG
		&& st->st_size >= min_size
		&& (max_size > 0 ? st->st_size <= max_size : 1)
		&& checksum_calc(&checksum, name) == 0 ))
	return;

    if (mode & MODE_TEMPDB) {

	if (csdb_exists(db, &checksum, sizeof checksum)) {

	    if (mode & MODE_REPORT)
		printf("%s\n", name);
	    if (mode & MODE_DELETE)
		unlink(name);

	} else {
	    csdb_store(db, &checksum, sizeof checksum);
	}

    } else if (mode & MODE_WRITEDB) {

	if (mode & MODE_REPORT)
	    printf("%s\n", name);

	csdb_store(db, &checksum, sizeof checksum);

    } else if (mode & MODE_READDB
	    && csdb_exists(db, &checksum, sizeof checksum)) {

	if (mode & MODE_REPORT)
	    printf("%s\n", name);

	if (mode & MODE_DELETE)
	    unlink(name);

    }
}

static void usage(void)
{
    printf(
"Usage: dupemap [OPTIONS] OPERATION PATH...\n"
"Where OPERATION is one of the operations listed in the manpage.\n"
"\n"
"Options:\n"
"  -d DATABASE   Read/write from a database on disk\n"
"  -I FILE       Read input file names from this file (\"-\" for stdin)\n"
"  -m MINSIZE    Exclude files below this size\n"
"  -M MAXSIZE    Exclude files above this size\n"
);
}

int main(int argc, char **argv)
{
    struct recur *recur;
    int c;
    char name[PATH_MAX], *dbname = NULL;
    char *file_names_from = NULL;
    struct stat st;
    dbhandle db;

    while ((c = getopt(argc, argv, "d:m:I:M:")) >= 0) {
	switch (c) {
	case 'd':
	    dbname = optarg;

	break; case 'I':
	    file_names_from = optarg;
	    
	break; case 'm':
#ifdef HAVE_ATOLL
	    min_size = (off_t)atoll(optarg);
#else
	    min_size = (off_t)atol(optarg);
#endif
	    if (min_size <= 0)
		min_size = 1;

	break; case 'M':
#ifdef HAVE_ATOLL
	    max_size = (off_t)atoll(optarg);
#else
	    max_size = (off_t)atol(optarg);
#endif

	break; default:
	    fprintf(stderr, "Error parsing options.\n");
	    usage();
	    return 1;
	}
    }

    argv += optind;
    argc -= optind;

    if (argc < 1) {
	usage();
	return 1;
    }

    mode = parse_mode(argv[0], dbname != NULL);
    if (!mode) {
	fprintf(stderr, "Invalid operation\n");
	return 1;
    }

    if (dbname) {
	if (csdb_open(&db, dbname, mode) != 0) {
	    fprintf(stderr, "Cannot open database\n");
	    return 1;
	}
    } else {
	if (csdb_open_tempdb(&db) != 0) {
	    fprintf(stderr, "Cannot open temporary database\n");
	    return 1;
	}
    }

    global_dbhandle = &db;

    signal(SIGINT,  signal_beforedeath);
    signal(SIGTERM, signal_beforedeath);
    signal(SIGSEGV, signal_beforedeath);
    signal(SIGPIPE, signal_beforedeath);
    
    if (argc > 1) {
	recur = recur_open(argv + 1);
	if (!recur) {
	    csdb_close(&db);
	    return 0;
	}

	while (recur_next(recur, name, &st) == 0) {
	    scan_file(name, &st, &db);
	}
	recur_close(recur);
    }

    if (file_names_from) {
	FILE *fh = strcmp(file_names_from, "-") == 0 ? stdin : 
		fopen(file_names_from, "r");
	
	if (!fh) {
	    fprintf(stderr, "Opening %s: %s\n",
		    file_names_from, strerror(errno));
	    csdb_close(&db);
	    return 1;
	}

	while (fgets(name, sizeof name, fh)) {
	    name[strlen(name)-1] = '\0'; /* kill newline */
	    if (stat(name, &st) == 0)
		scan_file(name, &st, &db);
	}
	fclose(fh);
    }

    csdb_close(&db);
    global_dbhandle = NULL;
    
    return 0;
}

#else /* not HAVE_DBM */
# include <stdio.h>

int main()
{
    printf(
"dupemap was not compiled because no ndbm.h was found on your system.  Please\n"
"install the development packages for Berkeley DB or GDBM and recompile.\n"
	  );
    return 0;
}
#endif /* not HAVE_DBM */
